File: ExtraWork/src/ticTacToe/test_board.py
import unittest
import board

class TestBoard(unittest.TestCase):
    def setUp(self):
        self.game = board.Board()

    def test_can_make_a_move(self):
        self.assertTrue(self.game.make_move(0, 0, 'X'))

    def test_cannot_move_to_same_position_twice(self):
        self.game.make_move(0, 0, 'X')
        with self.assertRaises(ValueError):
            self.game.make_move(0, 0, 'O')

    def test_row_winner(self):
        self.game.make_move(0, 0, 'X')
        self.game.make_move(0, 1, 'X')
        self.game.make_move(0, 2, 'X')
        self.assertEqual("Player X wins!", self.game.check_winner())

    def test_column_winner(self):
        self.game.make_move(0, 0, 'X')
        self.game.make_move(1, 0, 'X')
        self.game.make_move(2, 0, 'X')
        self.assertEqual("Player X wins!", self.game.check_winner())

    def test_diagonal_winner(self):
        self.game.make_move(0, 0, 'X')
        self.game.make_move(1, 1, 'X')
        self.game.make_move(2, 2, 'X')
        self.assertEqual("Player X wins!", self.game.check_winner())

    def test_other_diagonal_winner(self):
        self.game.make_move(0, 2, 'X')
        self.game.make_move(1, 1, 'X')
        self.game.make_move(2, 0, 'X')
        self.assertEqual("Player X wins!", self.game.check_winner())

    def test_draw(self):
        self.game.make_move(0, 0, 'X')
        self.game.make_move(1, 0, 'O')
        self.game.make_move(2, 0, 'X')
        self.game.make_move(1, 1, 'O')
        self.game.make_move(1, 2, 'X')
        self.game.make_move(2, 1, 'O')
        self.game.make_move(0, 1, 'X')
        self.game.make_move(0, 2, 'O')
        self.game.make_move(2, 2, 'X')
        self.assertEqual("It's a draw!", self.game.check_winner())

File: ExtraWork/test/TestBoard.java
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import ticTacToe.Board;

import static org.junit.jupiter.api.Assertions.*;

public class TestBoard {
    private Board game;

    @BeforeEach
    void setUp() {
        game = new Board();
    }
    @Test
    public void iCanMakeAMove() {
        assertTrue(game.makeMove(0,0,'X'));
    }

    @Test
    public void whenITryToMoveToSamePositionThatIHaveMovedToBefore() {
        game.makeMove(0,0,'X');
        assertThrows(IllegalArgumentException.class, () -> game.makeMove(0,0,'O'),
                "This Position has been occupied by player X");
    }

    @Test
    public void whenSameSymbolIsOnARowReturnThePlayerOfTheSymbolTheWinner() {
        game.makeMove(0, 0, 'X');
        game.makeMove(0, 1, 'X');
        game.makeMove(0, 2, 'X');
        String winner ="Player X wins!";
        assertEquals(winner, game.checkWinner());
    }

    @Test
    public void whenSameSymbolIsOnAColumnReturnThePlayerOfTheSymbolTheWinner() {
        game.makeMove(0, 0, 'X');
        game.makeMove(1, 0, 'X');
        game.makeMove(2, 0, 'X');
        String winner ="Player X wins!";
        assertEquals(winner, game.checkWinner());
    }

    @Test
    public void whenSameSymbolAreDiagonallyEqualReturnThePlayerOfTheSymbolTheWinner() {
        game.makeMove(0, 0, 'X');
        game.makeMove(1, 1, 'X');
        game.makeMove(2, 2, 'X');
        String winner ="Player X wins!";
        assertEquals(winner, game.checkWinner());
    }

    @Test
    public void whenSameSymbolAreDiagonallyEqualOnTheOtherSideReturnThePlayerOfTheSymbolTheWinner(){
        game.makeMove(0, 2, 'X');
        game.makeMove(1, 1, 'X');
        game.makeMove(2, 0, 'X');
        String winner ="Player X wins!";
        assertEquals(winner, game.checkWinner());
    }

    @Test
    public void whenBoardIsFullReturnDrawNoWinner() {
        game.makeMove(0, 0, 'X');
        game.makeMove(1, 0, 'O');
        game.makeMove(2, 0, 'X');
        game.makeMove(1, 1, 'O');
        game.makeMove(1, 2, 'X');
        game.makeMove(2, 1, 'O');
        game.makeMove(0,1,'X');
        game.makeMove(0,2,'O');
        game.makeMove(2,2,'X');
        String winner ="The game is over!" + " Draw!!!";
        assertEquals(winner, game.checkWinner());
    }
}
